keeper/components/orbit.go

```diff
@@ -0,0 +1,293 @@
+	// whether the protocol controller is paused or not.
+	PausedControllers collections.Map[int32, bool]
+}
+
+func NewOrbitComponent(
+	cdc codec.Codec,
+	sb *collections.SchemaBuilder,
+	logger log.Logger,
+	bankKeeper types.BankKeeperOrbit,
+) (*OrbitComponent, error) {
+	if logger == nil {
+		return nil, errors.New("logger cannot be nil")
+	}
+
+	orbitsKeeper := OrbitComponent{
+		logger:     logger.With(types.ComponentPrefix, types.OrbitsKeeperName),
+		bankKeeper: bankKeeper,
+
+		router: router.New[types.ProtocolID, interfaces.OrbitController](),
+
+		PausedOrbits: collections.NewMap(
+			sb,
+			types.PausedOrbitPrefix,
+			types.PausedOrbitsName,
+			collections.PairKeyCodec(collections.Int32Key, collections.StringKey),
+			collections.BoolValue,
+		),
+		PausedControllers: collections.NewMap(
+			sb,
+			types.PausedOrbitControllersPrefix,
+			types.PausedOrbitControllersName,
+			collections.Int32Key,
+			collections.BoolValue,
+		),
+	}
+
+	return &orbitsKeeper, orbitsKeeper.Validate()
+}
+
+func (k *OrbitComponent) Validate() error {
+	if k.logger == nil {
+		return errors.New("logger cannot be nil")
+	}
+	if k.bankKeeper == nil {
+		return errors.New("bank keeper cannot be nil")
+	}
+	if k.router == nil {
+		return errors.New("controllers router cannot be nil")
+	}
+	return nil
+}
+
+func (k *OrbitComponent) Logger() log.Logger {
+	return k.logger
+}
+
+func (k *OrbitComponent) Router() OrbitRouter {
+	return k.router
+}
+
+func (k *OrbitComponent) SetRouter(ocr OrbitRouter) {
+	if k.router != nil && k.router.Sealed() {
+		panic(errors.New("cannot reset a sealed controller router"))
+	}
+
+	k.router = ocr
+	k.router.Seal()
+}
+
+func (k *OrbitComponent) HandlePacket(ctx context.Context, packet *types.OrbitPacket) error {
+	if err := k.ValidatePacket(ctx, packet); err != nil {
+		return err
+	}
+
+	c, found := k.router.Route(packet.Orbit.ProtocolID())
+	if !found {
+		return errors.New("controller is not registered")
+	}
+
+	return c.HandlePacket(ctx, packet)
+}
+
+func (k *OrbitComponent) ValidatePacket(ctx context.Context, packet *types.OrbitPacket) error {
+	err := packet.Validate()
+	if err != nil {
+		return err
+	}
+
+	attr, err := packet.Orbit.CachedAttributes()
+	if err != nil {
+		return err
+	}
+
+	err = k.ValidateOrbit(ctx, packet.Orbit.ProtocolID(), attr.CounterpartyID())
+	if err != nil {
+		return err
+	}
+
+	return k.validateInitialConditions(ctx, packet)
+}
+
+func (k *OrbitComponent) ValidateOrbit(
+	ctx context.Context,
+	protocolID types.ProtocolID,
+	counterpartyID string,
+) error {
+	if err := k.validateController(ctx, protocolID); err != nil {
+		return err
+	}
+	return k.validateOrbit(ctx, protocolID, counterpartyID)
+}
+
+func (k *OrbitComponent) validateController(
+	ctx context.Context,
+	protocolID types.ProtocolID,
+) error {
+	isPaused, err := k.IsControllerPaused(ctx, protocolID)
+	if err != nil {
+		return err
+	}
+	if isPaused {
+		return errors.New("controller is paused")
```

Comment:
would be nice to specify in the error for which protocol id the controller is paused